feat(time-picker) - bug fixes and improvements

gavin-everett-genesys opened the pull request:
**Related Ticket:** https://inindca.atlassian.net/browse/COMUI-1437

Open to suggestions on how to further improve the time-picker input keyboard interactions.

A few things which have changed:

I tried to make it more functional like the native time input  although there is slight differences https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time
- I have added a leading `0` to the `12` hour time input before it just displayed the time as eg `9:00` instead of `09:00`. I am not sure was there a reason behind this to begin with ? (Could this be seen as a breaking change)
- Bug **(COMUI-1432)** should be resolved with this change so when typing in a digit the cursor will always start at the end of the input and not the current position of the cursor.
- Bug **(COMUI-1430)** should be resolved which is that in the 12hr hour input you cannot backspace if there is a leading 0 as this would be an invalid hour.

katie-bobbe-genesys commented:
> * I have added a leading `0` to the `12` hour time input before it just displayed the time as eg `9:00` instead of `09:00`. I am not sure was there a reason behind this to begin with ? (Could this be seen as a breaking change)

I think that it could have been an intentional decision to omit the leading zero from the 12 hour time input. What is the reason behind adding the leading `0`?

gavin-everett-genesys commented:
> > * I have added a leading `0` to the `12` hour time input before it just displayed the time as eg `9:00` instead of `09:00`. I am not sure was there a reason behind this to begin with ? (Could this be seen as a breaking change)
> 
> I think that it could have been an intentional decision to omit the leading zero from the 12 hour time input. What is the reason behind adding the leading `0`?

I think it was to make it more like the native input but as you said it may be intentional so best to leave it out.



> > * I have added a leading `0` to the `12` hour time input before it just displayed the time as eg `9:00` instead of `09:00`. I am not sure was there a reason behind this to begin with ? (Could this be seen as a breaking change)
> 
> I think that it could have been an intentional decision to omit the leading zero from the 12 hour time input. What is the reason behind adding the leading `0`?

I think I was trying to make it more like the native input but like you said it could have been an intentional decision to remove the `0` so probably best to leave it out of this PR.

katie-bobbe-genesys commented on src/components/beta/gux-time-picker/gux-time-picker.service.tsx:
```diff
@@ -140,23 +138,30 @@ export function getValidValueHourChange(
   clockType: GuxClockType,
   change: string,
-  selectionStart: number
+  hourInputEl: HTMLInputElement
```
I think it's confusing at first glance why this is being changed from `selectionStart` , which is a number, to `hourInputEl`, which is an HTML input element. I think the `getValidValueHourChange` function should probably not diverge too much from the the `getvValidValueMinuteChange` function, which takes `selectionStart` instead of an `inputElement`:
```
export function getValidValueMinuteChange(
  value: GuxISOHourMinute,
  change: string,
  selectionStart: number
): GuxISOHourMinute {
```
Maybe it would be clearer if `getValidValueHourChange` took `selectionStart:number` and `hourInputLength:number` as arguments instead of passing in `hourInputEl: HTMLInputElement`?
